pyiron_base/storage/datacontainer.py

```diff
@@ -851,62 +941,14 @@ def normalize_key(name):
 
             self.read_only = bool(hdf.get("READ_ONLY", False))
 
-    def nodes(self):
-        """
-        Iterator over keys to terminal nodes.
-
-        Returns:
-            :class:`list`: list of keys to normal values.
-        """
-        for k, v in self.items():
-            if not isinstance(v, DataContainer):
-                yield k
-
-    def _list_nodes(self):
-        return list(self.nodes())
-
-    def groups(self):
-        """
-        Iterate over keys to nested containers.
-
-        Returns:
-            :class:`list`: list of all keys to elements of :class:`DataContainer`.
-        """
-        for k, v in self.items():
-            if isinstance(v, DataContainer):
-                yield k
-
-    def _list_groups(self):
-        return list(self.groups())
-
-    @sentinel
-    def read(self, file_name, wrap=True):
-        """
-        Parse file as dictionary and add its keys to this container.
-
-        For supported file types, see :func:`.fileio.read`.
-
-        Errors during reading of the files generate a warning, but leave the container unchanged.
-
-        Args:
-            file_name(str): path to the input file
-            wrap(bool), if set to true will wrap the inputed data itself as a datacontainer inside the datacontainer
-
-        Raises:
-            :class:`ValueError`: if file extension doesn't match one of the supported ones
-        """
-        self.update(read(file_name), wrap=wrap)
-
-    def write(self, file_name):
-        """
-        Writes the DataContainer to a text file.
-
-        For supported file types, see :func:`.fileio.write`.
-
-        Args:
-            file_name(str): the name of the file to be writen to.
-        """
-        write(self.to_builtin(), file_name)
+    # HDFStub compat
+    def __getitem__(self, key):
+        value = super().__getitem__(key)
+        if not isinstance(value, HDFStub):
+            return value
+        else:
+            value = self[key] = value.load()
+            return value
```

Comment:
Once we have loaded a stub, we don't want to load it again a second time, so this overwrites the stub with the actual value.  I guess

```python
return self[key] := value.load()
```

would work as well, but is only available in newer python version (not sure which atm).

Reply:
ok it's a small thing, but my problem is that by returning `value` it might give the impression that it's the copy of the data, while in reality if you change it, it changes the content in `self[key]` as well. So I would simply set `self[key] = value.load()` and `return self[key]`.

Reply:
Won't work because we are inside `__getitem__`, so using `self[key]` will recurse without stopping.

Reply:
ah ok. Beautiful